dissect/executable/pe/helpers/resources.py

```diff
@@ -0,0 +1,418 @@
+
+if TYPE_CHECKING:
+    from dissect.cstruct.cstruct import cstruct
+    from dissect.cstruct.types.enum import EnumInstance
+
+    from dissect.executable.pe.helpers.sections import PESection
+    from dissect.executable.pe.pe import PE
+
+
+class ResourceManager:
+    """Base class to perform actions regarding the resources within the PE file.
+
+    Args:
+        pe: A `PE` object.
+        section: The section object that contains the resource table.
+    """
+
+    def __init__(self, pe: PE, section: PESection):
+        self.pe = pe
+        self.section = section
+        self.resources = OrderedDict()
+        self.raw_resources = []
+
+        self.parse_rsrc()
+
+    def parse_rsrc(self):
+        """Parse the resource directory entry of the PE file."""
+
+        rsrc_data = BytesIO(self.pe.read_image_directory(index=pestruct.IMAGE_DIRECTORY_ENTRY_RESOURCE))
+        self.resources = self._read_resource(rc_type="_root", data=rsrc_data, offset=0, level=1)
+
+    def _read_entries(self, data: bytes, directory: cstruct) -> list:
+        """Read the entries within the resource directory.
+
+        Args:
+            data: The data of the resource directory.
+            directory: The resource directory entry.
+
+        Returns:
+            A list containing the entries of the resource directory.
+        """
+
+        entries = []
+        for _ in range(0, directory.NumberOfNamedEntries + directory.NumberOfIdEntries):
+            entry_offset = data.tell()
+            entry = pestruct.IMAGE_RESOURCE_DIRECTORY_ENTRY(data)
+            self.raw_resources.append({"offset": entry_offset, "entry": entry, "data_offset": entry_offset})
+            entries.append(entry)
+        return entries
+
+    def _handle_data_entry(self, data: bytes, entry: cstruct, rc_type: str) -> Resource:
+        """Handle the data entry of a resource. This is the actual data associated with the directory entry.
+
+        Args:
+            data: The data of the resource.
+            entry: The resource directory entry.
+
+        Returns:
+            The resource that was given by name as a `Resource` object.
+        """
+
+        data.seek(entry.OffsetToDirectory)
+        data_entry = pestruct.IMAGE_RESOURCE_DATA_ENTRY(data)
+        self.pe.seek(data_entry.OffsetToData)
+        data = self.pe.read(data_entry.Size)
+        raw_offset = data_entry.OffsetToData - self.section.virtual_address
+        rsrc = Resource(
+            pe=self.pe,
+            section=self.section,
+            name=entry.Name,
+            entry_offset=entry.OffsetToData,
+            data_entry=data_entry,
+            rc_type=rc_type,
+        )
+        self.raw_resources.append(
+            {
+                "offset": entry.OffsetToDirectory,
+                "entry": data_entry,
+                "data": data,
+                "data_offset": raw_offset,
+                "resource": rsrc,
+            }
+        )
```

Comment:
This is kept for patching resources, adding them as a `dict` like proposed would need some more extensive refactoring of the code. We might opt for that indeed as it would be cleaner